Sources/KDTree.swift

```diff
@@ -62,27 +61,104 @@ extension KDTree {
             self = .node(left: .leaf, value: firstValue, dimension: currentSplittingDimension, right: .leaf)
         }
         else {
-            let sortedValues = values.sorted { (a, b) -> Bool in
-                return a.kdDimension(currentSplittingDimension) < b.kdDimension(currentSplittingDimension)
-            }
+            var median = values.startIndex + (values.endIndex - values.startIndex) / 2
             
-            var median = sortedValues.count / 2
-            let medianValue = sortedValues[median].kdDimension(currentSplittingDimension)
+            let selected = KDTree.quickSelect(n: median, arr: values, kdDimension: currentSplittingDimension)
+            let medianElement = selected[median]
+            let medianValue = medianElement.kdDimension(currentSplittingDimension)
             
             //Ensure left subtree contains currentSplittingDimension-coordinate strictly less than its parent node
             //Needed for 'contains' and 'removing' method.
-            while median >= 1 && abs(sortedValues[median-1].kdDimension(currentSplittingDimension) - medianValue) < Double.ulpOfOne {
+            while median >= 1 && median > selected.startIndex && abs(selected[median-1].kdDimension(currentSplittingDimension) - medianValue) < Double.ulpOfOne {
                 median -= 1
             }
             
-            let leftTree = KDTree(values: sortedValues[0..<median], depth: depth+1)
-            let rightTree = KDTree(values: sortedValues[median+1..<sortedValues.count], depth: depth+1)
+            let leftTree = KDTree(values: selected[selected.startIndex..<median], depth: depth+1)
+            let rightTree = KDTree(values: selected[median+1..<selected.endIndex], depth: depth+1)
             
-            self = .node(left: leftTree, value: sortedValues[median],
+            self = .node(left: leftTree, value: selected[median],
                          dimension: currentSplittingDimension, right: rightTree)
         }
     }
     
+    /*
+     Lomuto's partitioning algorithm.
+     This is conceptually simpler than Hoare's original scheme but less efficient.
+     The return value is the index of the pivot element in the new array. The left
+     partition is [low...p-1]; the right partition is [p+1...high], where p is the
+     return value.
+     The left partition includes all values smaller than the pivot, so
+     if the pivot value occurs more than once, its duplicates will be found in the
+     right partition.
+     */
+    private static func partitionLomuto(_ a: inout ArraySlice<Element>, kdDimension: Int) -> Int {
```

Comment:
Swift code styling defaults call for written out variables.
I have disabled the Swiftlint parameter `identifier_name` in this project , since I like to use variable names like `x` and `y` but in this case I would prefer `a` to be called `array`.